fl4health/clients/fed_pca_client.py

```diff
@@ -0,0 +1,144 @@
+
+    def get_parameters(self, config: Config) -> NDArrays:
+        assert self.model is not None and self.parameter_exchanger is not None
+        return self.parameter_exchanger.push_parameters(self.model, config=config)
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+        self.parameter_exchanger.pull_parameters(parameters, self.model, config)
+        self.save_model()
+
+    def narrow_config_type(self, config: Config, config_key: str, narrow_type_to: Type[T]) -> T:
+        if config_key not in config:
+            raise ValueError(f"{config_key} is not present in the Config.")
+
+        config_value = config[config_key]
+        if isinstance(config_value, narrow_type_to):
+            return config_value
+        else:
+            raise ValueError(f"Provided configuration key ({config_key}) value does not have correct type")
+
+    def get_parameter_exchanger(self, config: Config) -> FullParameterExchanger:
+        return FullParameterExchanger()
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        """
+        User defined method that returns a PyTorch Train DataLoader
+        and a PyTorch Validation DataLoader
+        """
+        raise NotImplementedError
+
+    def get_model(self, config: Config) -> PCAModule:
+        """
+        Returns an instance of the PCAModule.
+        """
+        low_rank = self.narrow_config_type(config, "low_rank", bool)
+        full_svd = self.narrow_config_type(config, "full_svd", bool)
+        rank_estimation = self.narrow_config_type(config, "rank_estimation", int)
+        return PCAModule(low_rank, full_svd, rank_estimation).to(self.device)
+
+    def setup_client(self, config: Config) -> None:
+        self.model = self.get_model(config).to(self.device)
+
+        train_loader, val_loader = self.get_data_loaders(config)
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+
+        self.parameter_exchanger = self.get_parameter_exchanger(config)
+
+        # The following lines are type ignored because torch datasets are not "Sized"
+        # IE __len__ is considered optionally defined. In practice, it is almost always defined
+        # and as such, we will make that assumption.
+        self.num_train_samples = len(self.train_loader.dataset)  # type: ignore
+        self.num_val_samples = len(self.val_loader.dataset)  # type: ignore
+
+        self.initialized = True
+
+    def get_data_tensor(self, data_loader: DataLoader) -> Tensor:
+        raise NotImplementedError
+
+    def fit(self, parameters: NDArrays, config: Config) -> Tuple[NDArrays, int, Dict[str, Scalar]]:
+        """Perform PCA using the locally held dataset."""
+        if not self.initialized:
+            self.setup_client(config)
+        center_data = self.narrow_config_type(config, "center_data", bool)
+        train_data_tensor = self.get_data_tensor(self.train_loader)
```

Comment:
these two lines can be merged into a single one without a large loss in readability I think.
```
train_data_tensor = self.get_data_tensor(self.train_loader).to(self.device)
```